<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="2 Preprocessing | Autoproliferation analysis B cells" />
<meta property="og:type" content="book" />





<meta name="author" content="Richard Affolter, Martin Emons, Philip Hartout" />



<meta name="description" content="2 Preprocessing | Autoproliferation analysis B cells">

<title>2 Preprocessing | Autoproliferation analysis B cells</title>

<script src="libs/header-attrs-2.6/header-attrs.js"></script>
<script src="libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="libs/navigation-1.1/tabsets.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
</style>
</head>

<body>

<div class="container-fluid main-container">


<div class="row">
<div class="col-sm-12">
<div id="TOC">
<ul>
<li><a href="1-introduction.html#introduction"><span class="toc-section-number">1</span> Introduction</a></li>
<li><a href="2-preprocessing.html#preprocessing"><span class="toc-section-number">2</span> Preprocessing</a>
<ul>
<li><a href="2-preprocessing.html#meta-data-preparation"><span class="toc-section-number">2.0.1</span> Meta-data preparation</a></li>
<li><a href="2-preprocessing.html#demultiplexing"><span class="toc-section-number">2.0.2</span> Demultiplexing</a></li>
</ul></li>
<li><a href="3-bracer.html#bracer"><span class="toc-section-number">3</span> BraCeR</a>
<ul>
<li><a href="3-bracer.html#assemble"><span class="toc-section-number">3.0.1</span> Assemble</a></li>
<li><a href="3-bracer.html#summarise"><span class="toc-section-number">3.0.2</span> Summarise</a></li>
</ul></li>
</ul>
</div>
</div>
</div>
<div class="row">
<div class="col-sm-12">
<div id="preprocessing" class="section level1" number="2">
<h1><span class="header-section-number">2</span> Preprocessing</h1>
<div id="meta-data-preparation" class="section level3" number="2.0.1">
<h3><span class="header-section-number">2.0.1</span> Meta-data preparation</h3>
<p>The patient meta-data was provided in an Rds file and was prepared with the following commands:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="2-preprocessing.html#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span></code></pre></div>
<pre><code>## 
## Attaching package: &#39;dplyr&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:stats&#39;:
## 
##     filter, lag</code></pre>
<pre><code>## The following objects are masked from &#39;package:base&#39;:
## 
##     intersect, setdiff, setequal, union</code></pre>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="2-preprocessing.html#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SummarizedExperiment)</span></code></pre></div>
<pre><code>## Loading required package: MatrixGenerics</code></pre>
<pre><code>## Loading required package: matrixStats</code></pre>
<pre><code>## 
## Attaching package: &#39;matrixStats&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:dplyr&#39;:
## 
##     count</code></pre>
<pre><code>## 
## Attaching package: &#39;MatrixGenerics&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:matrixStats&#39;:
## 
##     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,
##     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,
##     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,
##     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,
##     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,
##     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,
##     colWeightedMeans, colWeightedMedians, colWeightedSds,
##     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,
##     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,
##     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,
##     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,
##     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,
##     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,
##     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,
##     rowWeightedSds, rowWeightedVars</code></pre>
<pre><code>## Loading required package: GenomicRanges</code></pre>
<pre><code>## Loading required package: stats4</code></pre>
<pre><code>## Loading required package: BiocGenerics</code></pre>
<pre><code>## Loading required package: parallel</code></pre>
<pre><code>## 
## Attaching package: &#39;BiocGenerics&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:parallel&#39;:
## 
##     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
##     clusterExport, clusterMap, parApply, parCapply, parLapply,
##     parLapplyLB, parRapply, parSapply, parSapplyLB</code></pre>
<pre><code>## The following objects are masked from &#39;package:dplyr&#39;:
## 
##     combine, intersect, setdiff, union</code></pre>
<pre><code>## The following objects are masked from &#39;package:stats&#39;:
## 
##     IQR, mad, sd, var, xtabs</code></pre>
<pre><code>## The following objects are masked from &#39;package:base&#39;:
## 
##     anyDuplicated, append, as.data.frame, basename, cbind, colnames,
##     dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,
##     grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,
##     order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,
##     rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,
##     union, unique, unsplit, which.max, which.min</code></pre>
<pre><code>## Loading required package: S4Vectors</code></pre>
<pre><code>## 
## Attaching package: &#39;S4Vectors&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:dplyr&#39;:
## 
##     first, rename</code></pre>
<pre><code>## The following object is masked from &#39;package:base&#39;:
## 
##     expand.grid</code></pre>
<pre><code>## Loading required package: IRanges</code></pre>
<pre><code>## 
## Attaching package: &#39;IRanges&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:dplyr&#39;:
## 
##     collapse, desc, slice</code></pre>
<pre><code>## Loading required package: GenomeInfoDb</code></pre>
<pre><code>## Loading required package: Biobase</code></pre>
<pre><code>## Welcome to Bioconductor
## 
##     Vignettes contain introductory material; view with
##     &#39;browseVignettes()&#39;. To cite Bioconductor, see
##     &#39;citation(&quot;Biobase&quot;)&#39;, and for packages &#39;citation(&quot;pkgname&quot;)&#39;.</code></pre>
<pre><code>## 
## Attaching package: &#39;Biobase&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:MatrixGenerics&#39;:
## 
##     rowMedians</code></pre>
<pre><code>## The following objects are masked from &#39;package:matrixStats&#39;:
## 
##     anyMissing, rowMedians</code></pre>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="2-preprocessing.html#cb34-1" aria-hidden="true" tabindex="-1"></a>experiment <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="st">&quot;/Users/martinemons/polybox/Universitaet/MSc_CBB/HS2020/Statistical Analysis of HSD/vdj-analysis-local/sce_B.rds&quot;</span>, <span class="at">refhook =</span> <span class="cn">NULL</span>)</span>
<span id="cb34-2"><a href="2-preprocessing.html#cb34-2" aria-hidden="true" tabindex="-1"></a>cellUMI2SampleType <span class="ot">=</span> <span class="fu">colData</span>(experiment) <span class="sc">%&gt;%</span> as.data.frame <span class="sc">%&gt;%</span> <span class="fu">select</span>(lib_id, patient_id, Sample.Type, Diagnosis) <span class="sc">%&gt;%</span> <span class="fu">unique</span>()</span></code></pre></div>
<pre><code>## Loading required package: SingleCellExperiment</code></pre>
</div>
<div id="demultiplexing" class="section level3" number="2.0.2">
<h3><span class="header-section-number">2.0.2</span> Demultiplexing</h3>
<p>The files presented were derived from 10x-genomics. Since the pipeline BraCeR requires single cells in separate files the 10x genomics data had to be demultiplexed. This was done using the following python script</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb36-1"><a href="2-preprocessing.html#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co">#!/usr/bin/env python</span></span>
<span id="cb36-2"><a href="2-preprocessing.html#cb36-2" aria-hidden="true" tabindex="-1"></a><span class="co"># -*- coding: utf-8 -*-</span></span>
<span id="cb36-3"><a href="2-preprocessing.html#cb36-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-4"><a href="2-preprocessing.html#cb36-4" aria-hidden="true" tabindex="-1"></a><span class="co">&quot;&quot;&quot;demultiplexer.py</span></span>
<span id="cb36-5"><a href="2-preprocessing.html#cb36-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-6"><a href="2-preprocessing.html#cb36-6" aria-hidden="true" tabindex="-1"></a><span class="co">This script demultiplex the data, i.e. it creates a separate file for each</span></span>
<span id="cb36-7"><a href="2-preprocessing.html#cb36-7" aria-hidden="true" tabindex="-1"></a><span class="co">cell containing the contigs.</span></span>
<span id="cb36-8"><a href="2-preprocessing.html#cb36-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-9"><a href="2-preprocessing.html#cb36-9" aria-hidden="true" tabindex="-1"></a><span class="co">From: https://github.com/Teichlab/bracer/issues/21</span></span>
<span id="cb36-10"><a href="2-preprocessing.html#cb36-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-11"><a href="2-preprocessing.html#cb36-11" aria-hidden="true" tabindex="-1"></a><span class="co">&quot;&quot;&quot;</span></span>
<span id="cb36-12"><a href="2-preprocessing.html#cb36-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-13"><a href="2-preprocessing.html#cb36-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-14"><a href="2-preprocessing.html#cb36-14" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb36-15"><a href="2-preprocessing.html#cb36-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-16"><a href="2-preprocessing.html#cb36-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-17"><a href="2-preprocessing.html#cb36-17" aria-hidden="true" tabindex="-1"></a>list_of_cells <span class="op">=</span> <span class="bu">list</span>()</span>
<span id="cb36-18"><a href="2-preprocessing.html#cb36-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-19"><a href="2-preprocessing.html#cb36-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-20"><a href="2-preprocessing.html#cb36-20" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> main():</span>
<span id="cb36-21"><a href="2-preprocessing.html#cb36-21" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> folder <span class="kw">in</span> os.listdir(<span class="st">&quot;../data/demultiplexed&quot;</span>):</span>
<span id="cb36-22"><a href="2-preprocessing.html#cb36-22" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> <span class="st">&quot;preprocessed&quot;</span> <span class="kw">in</span> folder:</span>
<span id="cb36-23"><a href="2-preprocessing.html#cb36-23" aria-hidden="true" tabindex="-1"></a>            os.removedirs(folder)</span>
<span id="cb36-24"><a href="2-preprocessing.html#cb36-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-25"><a href="2-preprocessing.html#cb36-25" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> folder <span class="kw">in</span> os.listdir(<span class="st">&quot;../data/raw&quot;</span>):</span>
<span id="cb36-26"><a href="2-preprocessing.html#cb36-26" aria-hidden="true" tabindex="-1"></a>        <span class="cf">with</span> <span class="bu">open</span>(</span>
<span id="cb36-27"><a href="2-preprocessing.html#cb36-27" aria-hidden="true" tabindex="-1"></a>            <span class="ss">f&quot;../data/raw/</span><span class="sc">{</span>folder<span class="sc">}</span><span class="ss">/outs/filtered_contig.fasta&quot;</span>, <span class="st">&quot;r&quot;</span></span>
<span id="cb36-28"><a href="2-preprocessing.html#cb36-28" aria-hidden="true" tabindex="-1"></a>        ) <span class="im">as</span> <span class="bu">input</span>:</span>
<span id="cb36-29"><a href="2-preprocessing.html#cb36-29" aria-hidden="true" tabindex="-1"></a>            <span class="cf">for</span> line <span class="kw">in</span> <span class="bu">input</span>:</span>
<span id="cb36-30"><a href="2-preprocessing.html#cb36-30" aria-hidden="true" tabindex="-1"></a>                <span class="cf">if</span> line.startswith(<span class="st">&quot;&gt;&quot;</span>):</span>
<span id="cb36-31"><a href="2-preprocessing.html#cb36-31" aria-hidden="true" tabindex="-1"></a>                    cell <span class="op">=</span> line.split(<span class="st">&quot;-1_contig&quot;</span>)[<span class="dv">0</span>][<span class="dv">1</span>:]</span>
<span id="cb36-32"><a href="2-preprocessing.html#cb36-32" aria-hidden="true" tabindex="-1"></a>                    list_of_cells.append(cell)</span>
<span id="cb36-33"><a href="2-preprocessing.html#cb36-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-34"><a href="2-preprocessing.html#cb36-34" aria-hidden="true" tabindex="-1"></a>                <span class="cf">if</span> <span class="kw">not</span> os.path.exists(</span>
<span id="cb36-35"><a href="2-preprocessing.html#cb36-35" aria-hidden="true" tabindex="-1"></a>                    <span class="ss">f&quot;../data/demultiplexed/</span><span class="sc">{</span>folder<span class="sc">}</span><span class="ss">-preprocessed/&quot;</span></span>
<span id="cb36-36"><a href="2-preprocessing.html#cb36-36" aria-hidden="true" tabindex="-1"></a>                ):</span>
<span id="cb36-37"><a href="2-preprocessing.html#cb36-37" aria-hidden="true" tabindex="-1"></a>                    os.mkdir(<span class="ss">f&quot;../data/demultiplexed/</span><span class="sc">{</span>folder<span class="sc">}</span><span class="ss">-preprocessed/&quot;</span>)</span>
<span id="cb36-38"><a href="2-preprocessing.html#cb36-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-39"><a href="2-preprocessing.html#cb36-39" aria-hidden="true" tabindex="-1"></a>                <span class="cf">with</span> <span class="bu">open</span>(</span>
<span id="cb36-40"><a href="2-preprocessing.html#cb36-40" aria-hidden="true" tabindex="-1"></a>                    <span class="ss">f&quot;../data/demultiplexed/</span><span class="sc">{</span>folder<span class="sc">}</span><span class="ss">-preprocessed/</span><span class="sc">{</span>cell<span class="sc">}</span><span class="ss">.fasta&quot;</span>,</span>
<span id="cb36-41"><a href="2-preprocessing.html#cb36-41" aria-hidden="true" tabindex="-1"></a>                    <span class="st">&quot;a&quot;</span>,</span>
<span id="cb36-42"><a href="2-preprocessing.html#cb36-42" aria-hidden="true" tabindex="-1"></a>                ) <span class="im">as</span> output:</span>
<span id="cb36-43"><a href="2-preprocessing.html#cb36-43" aria-hidden="true" tabindex="-1"></a>                    output.write(line)</span>
<span id="cb36-44"><a href="2-preprocessing.html#cb36-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-45"><a href="2-preprocessing.html#cb36-45" aria-hidden="true" tabindex="-1"></a>        <span class="cf">with</span> <span class="bu">open</span>(</span>
<span id="cb36-46"><a href="2-preprocessing.html#cb36-46" aria-hidden="true" tabindex="-1"></a>            <span class="ss">f&quot;../data/demultiplexed/</span><span class="sc">{</span>folder<span class="sc">}</span><span class="ss">-preprocessed/list_of_cells.txt&quot;</span>,</span>
<span id="cb36-47"><a href="2-preprocessing.html#cb36-47" aria-hidden="true" tabindex="-1"></a>            <span class="st">&quot;w&quot;</span>,</span>
<span id="cb36-48"><a href="2-preprocessing.html#cb36-48" aria-hidden="true" tabindex="-1"></a>        ) <span class="im">as</span> <span class="bu">file</span>:</span>
<span id="cb36-49"><a href="2-preprocessing.html#cb36-49" aria-hidden="true" tabindex="-1"></a>            <span class="bu">file</span>.write(<span class="bu">str</span>(list_of_cells))</span>
<span id="cb36-50"><a href="2-preprocessing.html#cb36-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-51"><a href="2-preprocessing.html#cb36-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-52"><a href="2-preprocessing.html#cb36-52" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="va">__name__</span> <span class="op">==</span> <span class="st">&quot;__main__&quot;</span>:</span>
<span id="cb36-53"><a href="2-preprocessing.html#cb36-53" aria-hidden="true" tabindex="-1"></a>    main()</span></code></pre></div>
</div>
</div>
<p style="text-align: center;">
<a href="1-introduction.html"><button class="btn btn-default">Previous</button></a>
<a href="3-bracer.html"><button class="btn btn-default">Next</button></a>
</p>
</div>
</div>


</div>

<script>

// add bootstrap table styles to pandoc tables
$(document).ready(function () {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
});

</script>

</body>
</html>
