<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="3 Preprocessing of raw Data | Autoproliferation Analysis of B Cells in the context of Multiple Sclerosis" />
<meta property="og:type" content="book" />





<meta name="author" content="Richard Affolter, Martin Emons, Philip Hartout" />


<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.2/MathJax.js?config=TeX-AMS_CHTML-full" type="text/javascript"></script>

<meta name="description" content="3 Preprocessing of raw Data | Autoproliferation Analysis of B Cells in the context of Multiple Sclerosis">

<title>3 Preprocessing of raw Data | Autoproliferation Analysis of B Cells in the context of Multiple Sclerosis</title>

<script src="libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="libs/bootstrap-3.3.5/css/simplex.min.css" rel="stylesheet" />
<script src="libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="libs/navigation-1.1/tabsets.js"></script>
<script src="libs/accessible-code-block-0.0.1/empty-anchor.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
</style>
</head>

<body>

<div class="container-fluid main-container">


<div class="row">
<div class="col-sm-12">
<div id="TOC">
<ul>
<li><a href="1-introduction.html#introduction"><span class="toc-section-number">1</span> Introduction</a></li>
<li class="has-sub"><a href="2-exploratory-data-analysis.html#exploratory-data-analysis"><span class="toc-section-number">2</span> Exploratory Data Analysis</a><ul>
<li><a href="2-exploratory-data-analysis.html#sample-level-descriptive-statistics"><span class="toc-section-number">2.1</span> Sample-level Descriptive Statistics</a></li>
<li><a href="2-exploratory-data-analysis.html#cell-level-descriptive-statistics"><span class="toc-section-number">2.2</span> Cell-level Descriptive Statistics</a></li>
</ul></li>
<li class="has-sub"><a href="3-preprocessing-of-raw-data.html#preprocessing-of-raw-data"><span class="toc-section-number">3</span> Preprocessing of raw Data</a><ul>
<li><a href="3-preprocessing-of-raw-data.html#demultiplexing"><span class="toc-section-number">3.1</span> Demultiplexing</a></li>
</ul></li>
<li class="has-sub"><a href="4-bracer.html#bracer"><span class="toc-section-number">4</span> <code>BraCeR</code></a><ul>
<li><a href="4-bracer.html#assemble"><span class="toc-section-number">4.1</span> Assemble</a></li>
<li><a href="4-bracer.html#summarise"><span class="toc-section-number">4.2</span> Summarise</a></li>
<li><a href="4-bracer.html#packaging-for-downstream-analysis"><span class="toc-section-number">4.3</span> Packaging for downstream analysis</a></li>
</ul></li>
<li><a href="5-frequency-of-clonal-expansion.html#frequency-of-clonal-expansion"><span class="toc-section-number">5</span> Frequency of clonal expansion</a></li>
<li><a href="6-isotype-analysis.html#isotype-analysis"><span class="toc-section-number">6</span> Isotype Analysis</a></li>
<li><a href="7-discussion.html#discussion"><span class="toc-section-number">7</span> Discussion</a></li>
<li><a href="8-summary.html#summary"><span class="toc-section-number">8</span> Summary</a></li>
<li><a href="references.html#references">References</a></li>
</ul>
</div>
</div>
</div>
<div class="row">
<div class="col-sm-12">
<div id="preprocessing-of-raw-data" class="section level1">
<h1><span class="header-section-number">3</span> Preprocessing of raw Data</h1>
<div id="demultiplexing" class="section level2">
<h2><span class="header-section-number">3.1</span> Demultiplexing</h2>
<p>The files presented were derived from 10x-genomics. Since the pipeline BraCeR
requires single cells in separate files the 10x-genomics data had to be
demultiplexed, i.e.Â separate each cell data in a separate directory containing
the mapped sequences related to that cell. This was done using the following
python script, inspired by a maintainer of <code>BraCeR</code> on <a href="https://github.com/Teichlab/bracer/issues/21">GitHub</a>.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb14-1"><a href="3-preprocessing-of-raw-data.html#cb14-1"></a><span class="co">#!/usr/bin/env python</span></span>
<span id="cb14-2"><a href="3-preprocessing-of-raw-data.html#cb14-2"></a><span class="co"># -*- coding: utf-8 -*-</span></span>
<span id="cb14-3"><a href="3-preprocessing-of-raw-data.html#cb14-3"></a></span>
<span id="cb14-4"><a href="3-preprocessing-of-raw-data.html#cb14-4"></a><span class="co">&quot;&quot;&quot;demultiplexer.py</span></span>
<span id="cb14-5"><a href="3-preprocessing-of-raw-data.html#cb14-5"></a></span>
<span id="cb14-6"><a href="3-preprocessing-of-raw-data.html#cb14-6"></a><span class="co">This script demultiplex the data, i.e. it creates a separate file for each</span></span>
<span id="cb14-7"><a href="3-preprocessing-of-raw-data.html#cb14-7"></a><span class="co">cell containing the contigs.</span></span>
<span id="cb14-8"><a href="3-preprocessing-of-raw-data.html#cb14-8"></a></span>
<span id="cb14-9"><a href="3-preprocessing-of-raw-data.html#cb14-9"></a><span class="co">Adapted from: https://github.com/Teichlab/bracer/issues/21</span></span>
<span id="cb14-10"><a href="3-preprocessing-of-raw-data.html#cb14-10"></a></span>
<span id="cb14-11"><a href="3-preprocessing-of-raw-data.html#cb14-11"></a><span class="co">&quot;&quot;&quot;</span></span>
<span id="cb14-12"><a href="3-preprocessing-of-raw-data.html#cb14-12"></a></span>
<span id="cb14-13"><a href="3-preprocessing-of-raw-data.html#cb14-13"></a></span>
<span id="cb14-14"><a href="3-preprocessing-of-raw-data.html#cb14-14"></a><span class="im">import</span> os</span>
<span id="cb14-15"><a href="3-preprocessing-of-raw-data.html#cb14-15"></a></span>
<span id="cb14-16"><a href="3-preprocessing-of-raw-data.html#cb14-16"></a></span>
<span id="cb14-17"><a href="3-preprocessing-of-raw-data.html#cb14-17"></a>list_of_cells <span class="op">=</span> <span class="bu">list</span>()</span>
<span id="cb14-18"><a href="3-preprocessing-of-raw-data.html#cb14-18"></a></span>
<span id="cb14-19"><a href="3-preprocessing-of-raw-data.html#cb14-19"></a></span>
<span id="cb14-20"><a href="3-preprocessing-of-raw-data.html#cb14-20"></a><span class="kw">def</span> main():</span>
<span id="cb14-21"><a href="3-preprocessing-of-raw-data.html#cb14-21"></a>    <span class="cf">for</span> folder <span class="kw">in</span> os.listdir(<span class="st">&quot;../data/demultiplexed&quot;</span>):</span>
<span id="cb14-22"><a href="3-preprocessing-of-raw-data.html#cb14-22"></a>        <span class="cf">if</span> <span class="st">&quot;preprocessed&quot;</span> <span class="kw">in</span> folder:</span>
<span id="cb14-23"><a href="3-preprocessing-of-raw-data.html#cb14-23"></a>            os.removedirs(folder)</span>
<span id="cb14-24"><a href="3-preprocessing-of-raw-data.html#cb14-24"></a></span>
<span id="cb14-25"><a href="3-preprocessing-of-raw-data.html#cb14-25"></a>    <span class="cf">for</span> folder <span class="kw">in</span> os.listdir(<span class="st">&quot;../data/raw&quot;</span>):</span>
<span id="cb14-26"><a href="3-preprocessing-of-raw-data.html#cb14-26"></a>        <span class="cf">with</span> <span class="bu">open</span>(</span>
<span id="cb14-27"><a href="3-preprocessing-of-raw-data.html#cb14-27"></a>            <span class="ss">f&quot;../data/raw/</span><span class="sc">{</span>folder<span class="sc">}</span><span class="ss">/outs/filtered_contig.fasta&quot;</span>, <span class="st">&quot;r&quot;</span></span>
<span id="cb14-28"><a href="3-preprocessing-of-raw-data.html#cb14-28"></a>        ) <span class="im">as</span> <span class="bu">input</span>:</span>
<span id="cb14-29"><a href="3-preprocessing-of-raw-data.html#cb14-29"></a>            <span class="cf">for</span> line <span class="kw">in</span> <span class="bu">input</span>:</span>
<span id="cb14-30"><a href="3-preprocessing-of-raw-data.html#cb14-30"></a>                <span class="cf">if</span> line.startswith(<span class="st">&quot;&gt;&quot;</span>):</span>
<span id="cb14-31"><a href="3-preprocessing-of-raw-data.html#cb14-31"></a>                    cell <span class="op">=</span> line.split(<span class="st">&quot;-1_contig&quot;</span>)[<span class="dv">0</span>][<span class="dv">1</span>:]</span>
<span id="cb14-32"><a href="3-preprocessing-of-raw-data.html#cb14-32"></a>                    list_of_cells.append(cell)</span>
<span id="cb14-33"><a href="3-preprocessing-of-raw-data.html#cb14-33"></a></span>
<span id="cb14-34"><a href="3-preprocessing-of-raw-data.html#cb14-34"></a>                <span class="cf">if</span> <span class="kw">not</span> os.path.exists(</span>
<span id="cb14-35"><a href="3-preprocessing-of-raw-data.html#cb14-35"></a>                    <span class="ss">f&quot;../data/demultiplexed/</span><span class="sc">{</span>folder<span class="sc">}</span><span class="ss">-preprocessed/&quot;</span></span>
<span id="cb14-36"><a href="3-preprocessing-of-raw-data.html#cb14-36"></a>                ):</span>
<span id="cb14-37"><a href="3-preprocessing-of-raw-data.html#cb14-37"></a>                    os.mkdir(<span class="ss">f&quot;../data/demultiplexed/</span><span class="sc">{</span>folder<span class="sc">}</span><span class="ss">-preprocessed/&quot;</span>)</span>
<span id="cb14-38"><a href="3-preprocessing-of-raw-data.html#cb14-38"></a></span>
<span id="cb14-39"><a href="3-preprocessing-of-raw-data.html#cb14-39"></a>                <span class="cf">with</span> <span class="bu">open</span>(</span>
<span id="cb14-40"><a href="3-preprocessing-of-raw-data.html#cb14-40"></a>                    <span class="ss">f&quot;../data/demultiplexed/</span><span class="sc">{</span>folder<span class="sc">}</span><span class="ss">-preprocessed/</span><span class="sc">{</span>cell<span class="sc">}</span><span class="ss">.fasta&quot;</span>,</span>
<span id="cb14-41"><a href="3-preprocessing-of-raw-data.html#cb14-41"></a>                    <span class="st">&quot;a&quot;</span>,</span>
<span id="cb14-42"><a href="3-preprocessing-of-raw-data.html#cb14-42"></a>                ) <span class="im">as</span> output:</span>
<span id="cb14-43"><a href="3-preprocessing-of-raw-data.html#cb14-43"></a>                    output.write(line)</span>
<span id="cb14-44"><a href="3-preprocessing-of-raw-data.html#cb14-44"></a></span>
<span id="cb14-45"><a href="3-preprocessing-of-raw-data.html#cb14-45"></a>        <span class="cf">with</span> <span class="bu">open</span>(</span>
<span id="cb14-46"><a href="3-preprocessing-of-raw-data.html#cb14-46"></a>            <span class="ss">f&quot;../data/demultiplexed/</span><span class="sc">{</span>folder<span class="sc">}</span><span class="ss">-preprocessed/list_of_cells.txt&quot;</span>,</span>
<span id="cb14-47"><a href="3-preprocessing-of-raw-data.html#cb14-47"></a>            <span class="st">&quot;w&quot;</span>,</span>
<span id="cb14-48"><a href="3-preprocessing-of-raw-data.html#cb14-48"></a>        ) <span class="im">as</span> <span class="bu">file</span>:</span>
<span id="cb14-49"><a href="3-preprocessing-of-raw-data.html#cb14-49"></a>            <span class="bu">file</span>.write(<span class="bu">str</span>(list_of_cells))</span>
<span id="cb14-50"><a href="3-preprocessing-of-raw-data.html#cb14-50"></a></span>
<span id="cb14-51"><a href="3-preprocessing-of-raw-data.html#cb14-51"></a></span>
<span id="cb14-52"><a href="3-preprocessing-of-raw-data.html#cb14-52"></a><span class="cf">if</span> <span class="va">__name__</span> <span class="op">==</span> <span class="st">&quot;__main__&quot;</span>:</span>
<span id="cb14-53"><a href="3-preprocessing-of-raw-data.html#cb14-53"></a>    main()</span></code></pre></div>
<p>This script produces a folder structure with the following structure:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb15-1"><a href="3-preprocessing-of-raw-data.html#cb15-1"></a><span class="ex">demultiplexed/</span></span>
<span id="cb15-2"><a href="3-preprocessing-of-raw-data.html#cb15-2"></a>âââ <span class="ex">IJ009B_POOL1-preprocessed</span>   <span class="co"># Sample ID</span></span>
<span id="cb15-3"><a href="3-preprocessing-of-raw-data.html#cb15-3"></a>âÂ Â  âââ <span class="ex">AAACCTGAGAAGGGTA.fasta</span>  <span class="co"># Fasta file containing output of 10x genomics for one cell.</span></span>
<span id="cb15-4"><a href="3-preprocessing-of-raw-data.html#cb15-4"></a>âÂ Â  âââ <span class="ex">AAACCTGAGACTACAA.fasta</span></span>
<span id="cb15-5"><a href="3-preprocessing-of-raw-data.html#cb15-5"></a>âÂ Â  âââ <span class="ex">AAACCTGAGATACACA.fasta</span></span>
<span id="cb15-6"><a href="3-preprocessing-of-raw-data.html#cb15-6"></a>âÂ Â  âââ <span class="ex">AAACCTGCAAGTCTGT.fasta</span></span>
<span id="cb15-7"><a href="3-preprocessing-of-raw-data.html#cb15-7"></a>    âââ <span class="ex">AAACCTGCACGGTAGA.fasta</span></span>
<span id="cb15-8"><a href="3-preprocessing-of-raw-data.html#cb15-8"></a><span class="ex">...</span></span></code></pre></div>
</div>
</div>
<p style="text-align: center;">
<a href="2-exploratory-data-analysis.html"><button class="btn btn-default">Previous</button></a>
<a href="4-bracer.html"><button class="btn btn-default">Next</button></a>
</p>
</div>
</div>


</div>

<script>

// add bootstrap table styles to pandoc tables
$(document).ready(function () {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
});

</script>

</body>
</html>
