# Preprocessing

### Meta-data preparation

The patient meta-data was provided in an Rds file and was prepared with the following commands:
```{r}
library(dplyr)
library(SummarizedExperiment)
experiment = readRDS("/Users/martinemons/polybox/Universitaet/MSc_CBB/HS2020/Statistical Analysis of HSD/vdj-analysis-local/sce_B.rds", refhook = NULL)
cellUMI2SampleType = colData(experiment) %>% as.data.frame %>% select(lib_id, patient_id, Sample.Type, Diagnosis) %>% unique()
```

### Demultiplexing

The files presented were derived from 10x-genomics. Since the pipeline BraCeR requires single cells in separate files the 10x genomics data had to be demultiplexed. This was done using the following python script


```{r setup, include = FALSE}
library (reticulate)
use_python("usr/local/bin/python")
```


```{python, eval = FALSE}
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""demultiplexer.py

This script demultiplex the data, i.e. it creates a separate file for each
cell containing the contigs.

From: https://github.com/Teichlab/bracer/issues/21

"""


import os


list_of_cells = list()


def main():
    for folder in os.listdir("../data/demultiplexed"):
        if "preprocessed" in folder:
            os.removedirs(folder)

    for folder in os.listdir("../data/raw"):
        with open(
            f"../data/raw/{folder}/outs/filtered_contig.fasta", "r"
        ) as input:
            for line in input:
                if line.startswith(">"):
                    cell = line.split("-1_contig")[0][1:]
                    list_of_cells.append(cell)

                if not os.path.exists(
                    f"../data/demultiplexed/{folder}-preprocessed/"
                ):
                    os.mkdir(f"../data/demultiplexed/{folder}-preprocessed/")

                with open(
                    f"../data/demultiplexed/{folder}-preprocessed/{cell}.fasta",
                    "a",
                ) as output:
                    output.write(line)

        with open(
            f"../data/demultiplexed/{folder}-preprocessed/list_of_cells.txt",
            "w",
        ) as file:
            file.write(str(list_of_cells))


if __name__ == "__main__":
    main()
```

# BraCeR

With these demultiplexed cells one could run the BraCeR workflow with the two functionalities "assemble" and "summarise".

### Assemble

Assemble was performed by using the pre-assembled file from cellranger output and preparing the necessary files for the subsequent summarise step to function

```{python, eval = FALSE}
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""parallel_bracer.py

This executes multiple instances of docker in parallel to execute the bracer pipeline
"""

import subprocess
import os
import dotenv
from joblib import Parallel, delayed

# Load values from .env
DOTENV_KEY2VAL = dotenv.dotenv_values()

N_JOBS = -1
DOTENV_KEY2VAL["HOME_DIR"] = "/home/pjh/project-vdj-analysis"


def execute_docker_bracer(row, list_of_cells, patient):
    """Call docker in a subprocess for each cell

    Args:
        row (list): row containing the cell to be processed
        list_of_cells (list): list of cells containing all UMIs belonging to
        patients
        patient (str): patient to be processed
    Returns:
        None
    """
    cell = list_of_cells[row].split(".")[0:1]
    os.chdir(
        f"{DOTENV_KEY2VAL['HOME_DIR']}/data/demultiplexed/{patient}/{patient}-out"
    )
    subprocess.call(
        [
            "docker",
            "run",
            "--rm",
            "-v",
            f"$PWD:/scratch",
            "-w",
            "/scratch",
            "teichlab/bracer",
            "summarise",
            f"{DOTENV_KEY2VAL['HOME_DIR']}/data/demultiplexed/{patient}/{patient}-out",
        ]
    )


def files(path):
    list_of_files = list()
    for file in os.listdir(path):
        if os.path.isfile(os.path.join(path, file)):
            list_of_files.append(file)
    return list_of_files


def main():
    """Main function - lists patients and loop through patients. Each cell is
    processed in parallel using joblib

    Args:
        None

    Returns:
        None
    """
    list_of_patients = os.listdir(
        DOTENV_KEY2VAL["HOME_DIR"] + "/data/demultiplexed"
    )
    for patient in list_of_patients:
        list_of_cells = files(
            DOTENV_KEY2VAL["HOME_DIR"] + f"/data/demultiplexed/{patient}"
        )
        Parallel(n_jobs=N_JOBS, verbose=1)(
            delayed(execute_docker_bracer)(row, list_of_cells, patient)
            for row in range(len(list_of_cells))
        )


if __name__ == "__main__":
    main()
```

### Summarise

```{python, eval = FALSE}
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""parallel_bracer_summarise.py

This executes multiple instances of docker in parallel to execute the bracer summarise pipeline
"""

import subprocess
import os
import dotenv
from joblib import Parallel, delayed
from tqdm import tqdm


# Load values from .env
DOTENV_KEY2VAL = dotenv.dotenv_values()

N_JOBS = -1


def execute_docker_bracer(patient):
    """Call docker in a subprocess for each cell

    Args:
        row (list): row containing the cell to be processed
        list_of_cells (list): list of cells containing all UMIs belonging to
        patients
        patient (str): patient to be processed
    Returns:
        None
    """
    os.chdir(f"{DOTENV_KEY2VAL['HOME_DIR']}/data/demultiplexed/{patient}/")
    subprocess.call(
        [
            "docker",
            "run",
            "--rm",
            "-v",
            f"{DOTENV_KEY2VAL['HOME_DIR']}/data/demultiplexed/{patient}/:/scratch",
            "-w",
            "/scratch",
            "teichlab/bracer",
            "summarise",
            f"{patient}-out",
        ]
    )


def main():
    """Main function - lists patients and loop through patients. Each cell is
    processed in parallel using joblib

    Args:
        None

    Returns:
        None
    """
    list_of_patients = os.listdir(
        DOTENV_KEY2VAL["HOME_DIR"] + "/data/demultiplexed"
    )
    print(list_of_patients)
    Parallel(n_jobs=N_JOBS, verbose=1)(
        delayed(execute_docker_bracer)(row) for row in list_of_patients
    )


if __name__ == "__main__":
    main()
```

